integrations/tensorflow/bindings/python/pyiree/tf/support/tf_test_utils.py

```diff
@@ -429,12 +440,259 @@ def get_target_backends():
   return backends
 
 
+def _indent(input_str, indentation=2):
+  """Indents a string by the specified number of spaces, defaulting to 2."""
+  spaces = " " * indentation
+  lines = input_str.split("\n")
+  # Prepend spaces to each non-empty line.
+  lines = [f"{spaces}{line}" if len(line) else line for line in lines]
+  return "\n".join(lines)
+
+
+class ModuleCall:
+
+  def __init__(self, method_name, inputs, outputs, rtol=1e-6, atol=1e-6):
+    """Records the details of a call to a CompiledModule."""
+    self.method = method_name
+
+    # Deepcopy to safegard against mutation.
+    self.inputs = copy.deepcopy(inputs)
+    if outputs is not None:
+      outputs = copy.deepcopy(outputs)
+    else:
+      outputs = tuple()
+    self.outputs = outputs if isinstance(outputs, tuple) else (outputs,)
+
+    self.rtol = rtol
+    self.atol = atol
+
+  def get_tolerances(self):
+    """Gets the floating point tolerances associated with this call."""
+    return self.rtol, self.atol
+
+  def __str__(self):
+    prior_printoptions = np.get_printoptions()
+    np.set_printoptions(linewidth=NUMPY_LINEWIDTH)
+
+    header = f"Method: {self.method}"
+    inputs = "\n".join(_indent(str(value)) for value in self.inputs)
+    outputs = "\n".join(_indent(str(value)) for value in self.outputs)
+    tolerances = _indent(f"rtol={self.rtol}, atol={self.atol}")
+    body = f"Inputs:\n{inputs}\nOutputs:\n{outputs}\nTolerances:\n{tolerances}"
+    result = f"{header}\n{_indent(body)}"
+
+    np.set_printoptions(**prior_printoptions)
+    return result
+
+
+class TracedModule:
```

Comment:
Can you separate the TracedModule from the Trace itself? That is, TracedModule should only have functionality to wrap the CompiledModule, and it constructs a Trace.

Reply:
This sounds like a good idea. Do we care about trying to store the python function used to create the trace? We won't need it for benchmarking, and the user could copy the python source if they need it for some reason while debugging.

Reply:
I think that saving the python source is a good idea!

Reply:
Okay, now that we aren't passing the function to the constructor, we can get that data into the `Trace` via something like

```python
def forward_pass(module):
  ...
trace_name = forward_pass.__name__

llvm_info = tf_utils.BackendInfo.ALL['iree_llvmjit']
module = llvm_info.CompiledModule(SomeModule, llvm_info)
module = tf_test_utils.TracedModule(module, trace_name)

# Option 1.
forward_pass(module)
trace = module.get_trace()
trace.function = forward_pass

# Option 2.
def trace_function(self, function):  # On `TracedModule`
  self._trace.function = function
  function(self)

module.trace_function(foward_pass)
# now module.get_trace().function == forward_pass
```

What do you think?

Reply:
I don't think that storing the trace_function on the Trace makes sense. In my mind, Trace should be a serializable thing, and functions aren't meaningfully serializable. It sounds like what we want is for a Trace to hold a "provenance dictionary" (which we can, say, limit to a dictionary with string keys and values).

```
provenance = extract_provenance_from_function(trace_function) # extract name, file/line/col, and code snippet
# or 
# provenance = {"test_suite_name": ..., "test_method": ...}
trace = Trace(provenance_dict=provenance)
module = TracedModule(wrapped_module, trace_to_use=trace)
trace_function(module)
```

Logically, the tracing mechanics are actually independent from the trace to use and its provenance metadata.

Reply:
How about something like this? The trace also should have the info about the module and backend it's recording, so I wrote the `__init__` to do the metadata extraction of that and the function:

```python3
class Trace:
  
  def __init__(module, function):
    # Extract metadata from module and function.
    self.module_name = module.module_name
    self.backend = module.backend
    # Could be a dict or attrs of the Trace.
    self.function_name = function.__name__
    self.function_sourcefile = inspect.getsourcefile(function)
    self.function_lines = inspect.getsourcelines(function)
    self.function_source = inspect.getsource(function)
    # ... more info if desired.

    self.calls = []


def forward_pass(module)
  ...

module = IreeCompiledModule(...)
trace = Trace(module, forward_pass)
forward_pass(TracedModule(module, trace))
```

> Logically, the tracing mechanics are actually independent from the trace to use and its provenance metadata.

Yeah I think it's cleaner this way. The motivation for wrapping things together was to prevent the possibility of the function we say generates the trace, and what actually generated the trace from diverging.

Reply:
this looks perfect! Thanks for bearing with me on this :)